Feature/issue 1381 boardfarm in gitlab ci

odkq opened the pull request:
https://jira.prplfoundation.org/browse/PPM-30

odkq committed:
boardfarm-ci: Run boardfarm into docker-compose

test requirements and controller

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Use docker-compose from within boardfarm

Signed-off-by: odkq <pablo@odkq.com>

rmelotte committed:
boardfarm-ci: docker-compose.yml: use a variable for the volumes

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>
Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Add exit status check in _run_shell_cmd

Export $ROOT_DIR into images

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Preliminary version of prplmesh_compose.py

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Made the boardfarm container privileged

For dumpcap

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Use docker-compose run (not up)

With this, many instances can be launched. Each one get's the name
configured in the .json file, with the 'unique id' being the user.

Also, a volume is mapped where the tests expect the output log to
reside, from them to /tmp/user/beerocks/logs

Changed the entrypoint for boardfarm image to run the script and and
made a small shell script wrapper

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Fix dctest.sh --clean

Signed-off-by: odkq <pablo@odkq.com>

arnout commented on dctest.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+function dc {
```
`dc` is already an existing tool (decimal calculator, I think), so better call it something unique. "compose" for example.

arnout commented on dctest.sh:
```diff
@@ -0,0 +1,14 @@
+if [ $# -eq 0 ]; then
+    dc up boardfarm
+else
+    if [ $1 == "-h" ]; then
```
I believe shellcheck will complain here...
```suggestion
    if [ "$1" == "-h" ]; then
```

arnout commented on tools/docker/boardfarm-ci/docker-compose.yml:
```diff
@@ -0,0 +1,47 @@
+version: '3'
+services:
+    controller:
+        image: registry.gitlab.com/prpl-foundation/prplmesh/prplmesh-runner
```
We'll need an explicit tag here as well. Is there a way for docker-compose to get variables that get expanded? Like ARG in Dockerfile?

arnout commented on tools/docker/boardfarm-ci/docker-compose.yml:
```diff
@@ -0,0 +1,47 @@
+            - "8002"
+        volumes:
+            - "$ROOT_DIR:$ROOT_DIR"
+        entrypoint: ["/root/entrypoint.sh", "/usr/bin/start-controller-agent"]
```
OMG, this is a lot more readable than the stupid run script we have now... Nice!

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_compose.py:
```diff
@@ -0,0 +1,100 @@
+rootdir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../..'))
+
+
+class PrplMeshCompose(PrplMeshBase):
```
I don't think we should define a new class. Instead, change the PrplMeshDocker class to use compose instead of naked docker.

arnout reviewed (commented):
Not a full review - just a few loose comments. Please also see my comments in https://github.com/prplfoundation/prplMesh/pull/1492#pullrequestreview-439317396 where they're relevant for this one.

I don't really like this docker compose approach to launch several containers. Indeed, it forces us to duplicate the "knowledge" of which containers to use between the compose.yml file and the boardfarm config file. However, I do look how the compose yml file describes how to launch the container much better than how the shell script does it. So I'm all for using that rather than shell scripts to launch our containers. Wouldn't it be possible to make two compose.yml files, one for boardfarm itself and one for prplmesh, and do a compose boardfarm which launches boardfarm which launches two times compose prplmesh which luanches agent resp. controller?

odkq committed:
boardfarm-ci: Implementation of dctest in python 3.5

This will allow better extensibility when adding bridges and the rest

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
gitlab-ci: Check versions

Check versions of docker, docker-compose and python

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: add --shell parameter

odkq committed:
boardfarm-ci: Use curret user for volumes

odkq committed:
boardfarm-ci: Merge branch 'boardfarm-with-rax40' into feature/issue-1381-boardfarm-in-gitlab-ci

odkq committed:
boardfarm-ci: Merge branch 'boardfarm-with-rax40' into feature/issue-1381-boardfarm-in-gitlab-ci

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Remove existing network

This is functionality that do not belong on this branch

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Do not forward UID

To execute docker-compose inside boardfarm as root (again)
Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Remove user: references, run as root again

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
Enable running a different entrypoint

This allows running an entrypoint script from the code, mounted as
volume, instead of running the one in the uploaded image

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
Do not create the regular user into the container

As it is run as root (again) for now

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Prepopulate log directories

This way, even if the files belong to root, they can be removed by
gitlab-ci

Signed-off-by: odkq <pablo@odkq.com>

rmelotte committed:
docker: boardfarm-ci: add boardfarm to the docker image

The new version also need 2 new python packages.

Ideally, the requirements.txt file would be taken from the github
repository directly.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tests: boardfarm: use bft from the environment

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
rootdir: change to use the prplMesh directory itself

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

odkq committed:
boardfarm-ci: Add --comp parameter

This allows calling any docker-compose command from dc-test. For
example:

   ./dctest.py --comp images
   ./dctest.py --comp ps

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Allow simultaneous executions

This allows simultaneus executions of dctest.py, by using different
paths for logs and logs on devices based on an increasing id, when
running locally, and on the CI_PIPELINE_ID environment variable when run
in gitlab-ci.

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Allow simultaneous executions

This allows simultaneus executions of dctest.py, by using different
paths for logs and logs on devices based on an increasing id, when
running locally, and on the CI_PIPELINE_ID environment variable when run
in gitlab-ci.

odkq committed:
Merge branch 'feature/issue-1381-boardfarm-in-gitlab-ci' of https://github.com/prplfoundation/prplMesh into feature/issue-1381-boardfarm-in-gitlab-ci

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
Make the boardfarm instance itself parallel

The boardfarm container itself is able to run in parallel by using the
id passed as parameter or detected from the environment

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Added dctest to .gitlab-ci.yml

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Added dctest to .gitlab-ci.yml

odkq committed:
Merge branch 'feature/issue-1381-boardfarm-in-gitlab-ci' of https://github.com/prplfoundation/prplMesh into feature/issue-1381-boardfarm-in-gitlab-ci

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Correct dependency on .run-test-dctest

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Temporary replace bcl_unit_tests

To test dctest.py

odkq committed:
boardfarm-ci: Temporary replace bcl_unit_tests

To test dctest.py

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
Merge branch 'feature/issue-1381-boardfarm-in-gitlab-ci' of https://github.com/prplfoundation/prplMesh into feature/issue-1381-boardfarm-in-gitlab-ci

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Added correct tag in .gitlab-ci.yml

added tag 'boardfarm-compose' to dctest_one_test target

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
boardfarm-ci: Return != 0 when inner process do so

This allows for gitlab-ci to report a failing test

Signed-off-by: odkq <pablo@odkq.com>

odkq committed:
Use $CI_PIPELINE_ID tag for runner images.

Use 'latest' tag when testing locally (not in gitlab-ci)

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
boardfarm-ci: Forward the CI_PIPELINE_ID var

Forward the variable so it is used to tag the images for the runner and
agent inside boardfarm container

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
boardfarm-ci: Change owner of files to gitlab-runner

After running dctest.py, make sure the files created while building
boardfarm (as root) can be removed later by gitlab-runner

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
boardfarm-ci: Change ownership of the files

before exiting the script.

Adding it in .gitlab-ci.yml did not work as when the script fails next
lines are not executed

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
boardfarm-ci: Change destination for log directory

When it runs inside gitlab-ci, the directory is /tmp/beerocks/logs, not
/tmp/<user>/beerocks/logs

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
boardfarm-ci: missing import

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
boardfarm-ci: Set the ROOT_DIR and INSTALL_DIR

To /builds/prpl-foundation/prplMesh which is the one expected by the
runner and binaries built

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
boardfarm-ci: Do not try to chown from .gitlab-ci.yml

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
boardfarm-ci: Run boardfarm as a regular user

Setting the current uid and the group id of docker (to be able to access
the docker socket), this makes all the files created by the boardfarm
installation deletable by the gitlab-ci user

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
Merge branch 'master' into feature/issue-1381-boardfarm-in-gitlab-ci

odkq committed:
boardfarm-ci: Fixed 'missing device' bug

This appeared after merging from master

Signed-off-by: pablo <pablo@odkq.com>

rmelotte commented:
@odkq can you please:
- fix the shellcheck and flake8 issues that are reported on Gitlab: https://gitlab.com/prpl-foundation/prplMesh/-/pipelines/173659597
- make sure your "Signed-off-by" is there for every commit

Then we will have to cleanup the history (at least the easy parts), merge it, then do additional cleanup like removing the old boardfarm test, and integrating https://github.com/prplfoundation/prplMesh/pull/1556.

odkq committed:
boardfarm-ci: Removed unused shell script

Alternative entrypoint for the runner image when running as a regular user

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
boardfarm-ci: Fix device parameter passing

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
boardfarm-ci: Shellcheck fixes

Signed-off-by: pablo <pablo@odkq.com>

odkq committed:
boardfarm-ci: Fixes for PEP-8 compliance

Signed-off-by: pablo <pablo@odkq.com>

rmelotte commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_compose.py:
```diff
@@ -0,0 +1,112 @@
+                                 "run", "start-controller")
+            time.sleep(self.delay)
+            self.controller_entity = \
+                ALEntityDocker(self.name, is_controller=True, compose=True)
```
```suggestion
                ALEntityDocker(self.name, device=self, is_controller=True, compose=True)
```

rmelotte commented:
Superseded by https://github.com/prplfoundation/prplMesh/pull/1587
